aiida/orm/querybuilder.py

```diff
@@ -989,20 +993,36 @@ def _get_aiida_entity_res(value) -> Any:
         except TypeError:
             return value
 
-    def first(self) -> Optional[List[Any]]:
-        """Executes the query, asking for the first row of results.
+    @overload
+    def first(self, flat: Literal[False]) -> Optional[list[Any]]:
+        ...
+
+    @overload
+    def first(self, flat: Literal[True]) -> Optional[Any]:
+        ...
+
+    def first(self, flat: bool = False) -> Optional[list[Any] | Any]:
```

Comment:
```suggestion
    def first(self, flat: bool = False) -> None | list[Any] | Any:
```

Will approve anyway, but you are kind of mixing old and new type annotations in these changes